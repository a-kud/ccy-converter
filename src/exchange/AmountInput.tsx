import React, { useState } from "react";
import styled from "styled-components";
interface AmountInputProps {
  value: string;
  inputRef: React.RefObject<HTMLInputElement>;
  onChange: (text: string) => void;
  className: string;
  placeholder: string;
}

const Input = styled("input")`
  &:focus {
    outline: none;
  }
`;
Input.displayName = "AmountInput";

export function AmountInput({
  value,
  inputRef,
  onChange,
  className,
  placeholder,
}: AmountInputProps) {
  const measureRef = React.useRef<HTMLDivElement>(null);
  const [inputWidth, setInputWidth] = useState(0);

  const handleChange = (e) => {
    onChange(e.target.value);
  };

  React.useEffect(() => {
    const { current: measureEl } = measureRef;
    if (measureEl) {
      measureEl.textContent = value;
      setInputWidth(measureEl.offsetWidth);
    }
  }, [value, inputRef, setInputWidth]);

  return (
    <Input
      dir="rtl"
      className={className}
      ref={inputRef}
      onChange={handleChange}
      value={value}
      width={inputWidth}
      placeholder={placeholder}
    />
  );
}
