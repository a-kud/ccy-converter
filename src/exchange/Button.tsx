import React from "react";

interface ButtonProps {
  children: React.ReactChild;
  onClick?: () => void;
  className?: string;
}

export function Button({ className, children, onClick }: ButtonProps) {
  const handleClick = () => {
    if (onClick) {
      onClick();
    }
  };

  return (
    <button className={className} disabled={!onClick} onClick={handleClick}>
      {children}
    </button>
  );
}
