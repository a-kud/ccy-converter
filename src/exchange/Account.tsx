import React from "react";
import styled from "styled-components";
import { color } from "../styles";
import { ExchangeRateType } from "../types";
import { AmountInput } from "./AmountInput";
import { Button } from "./Button";
import { TotalAmount } from "./TotalAmount";
import { ExchangeRate } from "./ExchangeRate";

export const TotalContainer = styled("div")<{ invalid?: boolean }>`
  color: ${(props) => (props.invalid ? color.red : color.whiteTransparent)};
`;

export interface AccountProps {
  currency: string;
  amount: string;
  totalAmount: string;
  base?: boolean;
  exchangeRate?: ExchangeRateType;
  invalid?: boolean;
  onAmountChange: (text: string) => void;
  onAccountChange: (next?: boolean) => void;
  placeholder: string;
}

export const Account = React.memo(
  ({
    base,
    currency,
    amount,
    totalAmount,
    exchangeRate,
    invalid,
    onAmountChange,
    onAccountChange,
    placeholder,
  }: AccountProps) => {
    const inputRef = React.useRef<HTMLInputElement>(null);
    const handleClick = () => {
      const { current } = inputRef;
      if (current) {
        current.focus();
      }
    };

    const handleNext = () => {
      onAccountChange(true);
    };

    const handlePrev = () => {
      onAccountChange();
    };

    return (
      <div className="text-white bg-dark py-3" onClick={handleClick}>
        <div className="d-flex align-items-baseline">
          <Button className="btn btn-dark bg-dark" onClick={handlePrev}>
            〈
          </Button>
          <p>{currency}</p>
          <Button className="btn btn-dark bg-dark" onClick={handleNext}>
            〉
          </Button>
          <div className="d-flex flex-grow-1 justify-content-end w-25">
            {amount.length > 0 && <span>{base ? "-" : "+"}</span>}
            <AmountInput
              className="w-50 border-0 text-light bg-dark"
              placeholder={placeholder}
              value={amount}
              onChange={onAmountChange}
              inputRef={inputRef}
            />
          </div>
        </div>
        <div className="d-flex">
          <TotalContainer invalid={invalid}>
            <TotalAmount currency={currency} amount={totalAmount} />
          </TotalContainer>
          {exchangeRate && (
            <div className="d-flex flex-grow-1 justify-content-end">
              <ExchangeRate data={exchangeRate} rounded />
            </div>
          )}
        </div>
      </div>
    );
  }
);
