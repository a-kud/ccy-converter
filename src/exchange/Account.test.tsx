import { shallow } from "enzyme";
import toJson from "enzyme-to-json";
import { Account } from "./Account";
import { AmountInput } from "./AmountInput";
import { Button } from "./Button";

describe("Account", () => {
  let wrapper;
  const mockFn = jest.fn();
  beforeEach(() => {
    wrapper = shallow(
      <Account
        placeholder=""
        amount="10"
        currency="EUR"
        totalAmount="120"
        onAmountChange={mockFn}
        onAccountChange={mockFn}
      />
    );
  });
  it("renders correctly", () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("fires onAmountChange", () => {
    const props = wrapper.find(AmountInput).props();
    props.onChange("changed");

    expect(mockFn).toBeCalledWith("changed");
  });

  it("fires onChangeAccount", () => {
    const nextButtonProps = wrapper.find(Button).at(1).props();
    expect(nextButtonProps.children).toStrictEqual("〉");

    nextButtonProps.onClick!();

    expect(mockFn).toBeCalledWith(true);
  });

  it("fires calls onChangeAccount", () => {
    const prevButtonProps = wrapper.find(Button).at(0).props();
    expect(prevButtonProps.children).toStrictEqual("〈");

    prevButtonProps.onClick!();

    expect(mockFn).toBeCalledWith();
  });
});
