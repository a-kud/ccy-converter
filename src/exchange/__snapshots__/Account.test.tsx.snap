// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Account renders correctly 1`] = `
<div
  className="text-white bg-dark py-3"
  onClick={[Function]}
>
  <div
    className="d-flex align-items-baseline"
  >
    <Button
      className="btn btn-dark bg-dark"
      onClick={[Function]}
    >
      〈
    </Button>
    <p>
      EUR
    </p>
    <Button
      className="btn btn-dark bg-dark"
      onClick={[Function]}
    >
      〉
    </Button>
    <div
      className="d-flex flex-grow-1 justify-content-end w-25"
    >
      <span>
        +
      </span>
      <AmountInput
        className="w-50 border-0 text-light bg-dark"
        inputRef={
          Object {
            "current": null,
          }
        }
        onChange={[MockFunction]}
        placeholder=""
        value="10"
      />
    </div>
  </div>
  <div
    className="d-flex"
  >
    <styled.div>
      <TotalAmount
        amount="120"
        currency="EUR"
      />
    </styled.div>
  </div>
</div>
`;
